### ISSUE TITLE
Compilation error on RedHat linux cluster.

### ISSUE THREAD

**Issue by caofan** [bazelbuild/bazel#2083]
Please provide the following information. The more we know about your system and use case, the more easily and likely we can help.

### Description of the problem / feature request / question:
I tried to compile bazel on a linux cluster but got the following error:

```
INFO: You can skip this first step by providing a path to the bazel binary as second argument:
INFO:    ./compile.sh compile /path/to/bazel
🍃  Building Bazel from scratch.......
🍃  Building Bazel with Bazel.
.WARNING: /tmp/bazel_MCS3hRKY/out/external/bazel_tools/WORKSPACE:1: Workspace name in /tmp/bazel_MCS3hRKY/out/external/bazel_tools/WORKSPACE (@io_bazel) does not match the name given in the repository's definition (@bazel_tools); this will cause a build error in future versions.
INFO: Found 1 target...
ERROR: /home/users/nus/csicf/bazel-0.4.0/src/main/tools/BUILD:3:1: C++ compilation of rule '//src/main/tools:process-wrapper' failed: process-wrapper failed: error executing command
  (cd /tmp/bazel_MCS3hRKY/out/bazel-sandbox/366c25e7-9ca7-4bca-822a-29e5d0801b05-6/execroot/bazel-0.4.0 && \
  exec env - \
    LD_LIBRARY_PATH=/app/gcc/SCILibs/lib64:/app/gcc/SCILibs/lib:/app/gnuplot/5.0.3/lib:/app/gcc/4.9.3/lib64:/app/gcc/4.9.3/lib:/app/intel/compilers_and_libraries_2016.1.150/linux/mkl/lib/intel64:/app/isl/0.14/lib:/app/mpc/1.0.3/lib:/app/mpfr/3.1.4/lib:/app/gmp/6.1.0/lib:/app/src/ncbi-blast-2.4.0+-src/c++/local/ncbi-vdb-2.6.1/lib64: \
    PATH=/app/gnuplot/5.0.3/bin:/app/gcc/4.9.3/bin:/app/binutils/2.26/bin:/app/java/default/bin:/home/users/nus/csicf/Basset/src:/home/users/nus/csicf/anaconda3/bin:/home/users/nus/csicf/lib/git/bin:/usr/lib64/qt-3.3/bin:/app/pbs/bin:/opt/pbs/default/sbin:/opt/pbs/default/bin:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/sbin:/opt/ibutils/bin:/sbin:/usr/sbin:/home/users/nus/csicf/bin \
  /tmp/bazel_MCS3hRKY/out/execroot/bazel-0.4.0/_bin/process-wrapper -1 5 - - /app/gcc/4.9.3/bin/gcc -U_FORTIFY_SOURCE '-D_FORTIFY_SOURCE=1' -fstack-protector -Wall -Wl,-z,-relro,-z,now -B/app/gcc/4.9.3/bin -B/usr/bin -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-canonical-system-headers -fno-omit-frame-pointer -g0 -O2 -DNDEBUG -ffunction-sections -fdata-sections -MD -MF bazel-out/local-opt/bin/src/main/tools/_objs/process-wrapper/src/main/tools/process-tools.d -iquote . -iquote bazel-out/local-opt/genfiles -iquote external/bazel_tools -iquote bazel-out/local-opt/genfiles/external/bazel_tools -isystem external/bazel_tools/tools/cpp/gcc3 '-std=c99' -fno-canonical-system-headers -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c src/main/tools/process-tools.c -o bazel-out/local-opt/bin/src/main/tools/_objs/process-wrapper/src/main/tools/process-tools.o).
/tmp/cc4vjcBH.s: Assembler messages:
/tmp/cc4vjcBH.s:682: Error: expecting string instruction after `rep'
Target //src:bazel failed to build
INFO: Elapsed time: 7.303s, Critical Path: 2.98s

ERROR: Could not build Bazel
```

################################################################
gcc 4.9.3 and binutils  2.26 were used. 
gcc is installed in /app/gcc/4.9.3 and binutils is installed in /app/binutils/2.26. 

The toolchain is :

```
toolchain {
  abi_version: "local"
  abi_libc_version: "local"
  builtin_sysroot: ""
  compiler: "compiler"
  host_system_name: "local"
  needsPic: true
  supports_gold_linker: false
  supports_incremental_linker: false
  supports_fission: false
  supports_interface_shared_objects: false
  supports_normalizing_ar: false
  supports_start_end_lib: false
  target_libc: "local"
  target_cpu: "local"
  target_system_name: "local"
  toolchain_identifier: "local_linux"

  tool_path { name: "ar" path: "/app/binutils/2.26/bin/ar" }
  tool_path { name: "compat-ld" path: "/app/binutils/2.26/bin/ld" }
  tool_path { name: "cpp" path: "/app/gcc/4.9.3/bin/cpp" }
  tool_path { name: "dwp" path: "/app/gcc/4.9.3/bin/dwp" }
  tool_path { name: "gcc" path: "/app/gcc/4.9.3/bin/gcc" }
  cxx_flag: "-std=c++0x"
  linker_flag: "-lstdc++"
  linker_flag: "-B/app/gcc/4.9.3/bin/"
  linker_flag: "-B/app/binutils/2.26/bin/"

  # TODO(bazel-team): In theory, the path here ought to exactly match the path
  # used by gcc. That works because bazel currently doesn't track files at
  # absolute locations and has no remote execution, yet. However, this will need
  # to be fixed, maybe with auto-detection?
  cxx_builtin_include_directory: "/app/gcc/4.9.3/lib/"
  cxx_builtin_include_directory: "/app/gcc/4.9.3/include"
  cxx_builtin_include_directory: "/app/binutils/2.26/lib/"
  cxx_builtin_include_directory: "/app/binutils/2.26/include/"
  tool_path { name: "gcov" path: "/usr/bin/gcov" }

  # C(++) compiles invoke the compiler (as that is the one knowing where
  # to find libraries), but we provide LD so other rules can invoke the linker.
  tool_path { name: "ld" path: "/app/binutils/2.26/bin/ld" }

  tool_path { name: "nm" path: "/app/binutils/2.26/bin/nm" }
  tool_path { name: "objcopy" path: "/app/binutils/2.26/bin/objcopy" }
  objcopy_embed_flag: "-I"
  objcopy_embed_flag: "binary"
  tool_path { name: "objdump" path: "/app/binutils/2.26/bin/objdump" }
  tool_path { name: "strip" path: "/app/binutils/2.26/bin/strip" }

  # Anticipated future default.
  unfiltered_cxx_flag: "-no-canonical-prefixes"
  unfiltered_cxx_flag: "-fno-canonical-system-headers"

  # Make C++ compilation deterministic. Use linkstamping instead of these
  # compiler symbols.
  unfiltered_cxx_flag: "-Wno-builtin-macro-redefined"
  unfiltered_cxx_flag: "-D__DATE__=\"redacted\""
  unfiltered_cxx_flag: "-D__TIMESTAMP__=\"redacted\""
  unfiltered_cxx_flag: "-D__TIME__=\"redacted\""

  # Security hardening on by default.
  # Conservative choice; -D_FORTIFY_SOURCE=2 may be unsafe in some cases.
# We need to undef it before redefining it as some distributions now have
  # it enabled by default.
  compiler_flag: "-U_FORTIFY_SOURCE"
  compiler_flag: "-D_FORTIFY_SOURCE=1"
  compiler_flag: "-fstack-protector"
  linker_flag: "-Wl,-z,relro,-z,now"

  # Enable coloring even if there's no attached terminal. Bazel removes the
  # escape sequences if --nocolor is specified. This isn't supported by gcc
  # on Ubuntu 14.04.
  # compiler_flag: "-fcolor-diagnostics"

  # All warnings are enabled. Maybe enable -Werror as well?
  compiler_flag: "-Wall"
  # Enable a few more warnings that aren't part of -Wall.
  compiler_flag: "-Wunused-but-set-parameter"
  # But disable some that are problematic.
  compiler_flag: "-Wno-free-nonheap-object" # has false positives

  # Keep stack frames for debugging, even in opt mode.
  compiler_flag: "-fno-omit-frame-pointer"

  # Anticipated future default.
  linker_flag: "-no-canonical-prefixes"
  # Have gcc return the exit code from ld.
  linker_flag: "-pass-exit-codes"
  # Stamp the binary with a unique identifier.
  linker_flag: "-Wl,--build-id=md5"
  linker_flag: "-Wl,--hash-style=gnu"
  # Gold linker only? Can we enable this by default?
  # linker_flag: "-Wl,--warn-execstack"
  # linker_flag: "-Wl,--detect-odr-violations"

compilation_mode_flags {
    mode: DBG
    # Enable debug symbols.
    compiler_flag: "-g"
  }
  compilation_mode_flags {
    mode: OPT

    # No debug symbols.
    # Maybe we should enable https://gcc.gnu.org/wiki/DebugFission for opt or
    # even generally? However, that can't happen here, as it requires special
    # handling in Bazel.
    compiler_flag: "-g0"

    # Conservative choice for -O
    # -O3 can increase binary size and even slow down the resulting binaries.
    # Profile first and / or use FDO if you need better performance than this.
    compiler_flag: "-O2"

    # Disable assertions
    compiler_flag: "-DNDEBUG"

    # Removal of unused code and data at link time (can this increase binary size in some cases?).
    compiler_flag: "-ffunction-sections"
    compiler_flag: "-fdata-sections"
    linker_flag: "-Wl,--gc-sections"
  }
  linking_mode_flags { mode: DYNAMIC }

```

### Environment info
* Operating System:
RedHat Linux,
uname -m:   x86_64

* Bazel version (output of `bazel info release`):
0.4.0

**Comment by damienmg** [comment]
`gcc -version`?

Have you tried the steps there: https://github.com/bazelbuild/bazel/wiki/FAQ#q2?

**Comment by damienmg** [comment]
oops I just saw the gcc version.

What does `which gcc` says? And try to do `export CC=/app/gcc/4.9.3/gcc; ./compile.sh`

**Comment by caofan** [comment]
`which gcc` points to `/app/gcc/4.9.3/bin/gcc`.
I have exported `CC=gcc` in .bashrc.

I just tried to use the command line you gave, the same error still occurs.

**Comment by damienmg** [comment]
Where did you get the CROSSTOOL file from?

**Comment by caofan** [comment]
I edited the CROSSTOOL from the source tar ball.

**Comment by damienmg** [comment]
Do you know that this file is no longer used for Linux? What is the change you applied to this file?

**Comment by caofan** [comment]
I don't know about that...
I have changed the paths to the tools and added `linker_flag` and `cxx_builtin_include_directory` to include the `binutils`. I am not sure whether this is the correct way to do it.

**Comment by damienmg** [comment]
Ok I am guessing we cannot generate those flag:
```
  linker_flag: "-B/app/gcc/4.9.3/bin/"
  linker_flag: "-B/app/binutils/2.26/bin/"
```

Can you try to add them to https://github.com/bazelbuild/bazel/blob/master/tools/cpp/cc_configure.bzl#L223 this list?

**Comment by caofan** [comment]
Still no luck.

Actually I was able to get it compiled using linuxbrew's gcc programs and libraries earlier today.
But the problem with that was it gives 
`/home/users/nus/csicf/bin/bazel: /lib64/libc.so.6: version 'GLIBC_2.14' not found (required by /home/users/nus/csicf/bin/bazel)`
when I ran bazel.

Because Linuxbrew does not work well with LD_LIBRARY_PATH, I thought it might be better to go without it and stick to what the server provides.

I notice that the linking mode is dynamic, can the "GLIBC_2.14" issue be solved using a static compilation? I am very new to this, and probably misunderstood..

**Comment by damienmg** [comment]
Do you know how gcc was installed so we can repro that in a Docker container? Also which version of RHEL?

**Comment by caofan** [comment]
I think I figured out why it failed. 
System version is RedHat 6.6. The gcc and binutils were installed as modules which can be loaded through command `module load`. 
I checked the gcc `lib` directories and the `lib` of `binutils` and realised that the these modules do not have the glibc library `libc.so.6`. I think that's the reason why it failed as it still tries to compile using the system glibc-2.12.

I managed to run bazel with `LD_LIBRARY_PATH=/home/users/.linuxbrew/lib64:/home/users/.linuxbrew/lib:$LD_LIBRARY_PATH" /home/users/.linuxbrew/lib/ld-2.19.so bazel` from [http://stackoverflow.com/questions/33655731/error-while-importing-tensorflow-in-python2-7-in-ubuntu-12-04-glibc-2-17-not-f/34897674#34897674](url)

By setting the `LD_LIBRARY_PATH` alone without `/home/users/.linuxbrew/lib/ld-2.19.so` would result in segmentation fault.

Thanks for the help anyway.
